File: pattern/pattern.py
min_seq_length = 4
Seq = list()
length = [0] * 10


def middle(a, b):
    """Return the node between a and b if there is one, or None."""
    if (a + b) % 2 != 0:
        return None
    mid = (a + b) / 2
    if mid == 5 or a % 3 == b % 3 or (a - 1) / 3 == (b - 1) / 3:
        return mid


def next(base):
    if len(base) == 0:
        return list(xrange(1, 10))
    tmp = list()
    for i in xrange(1, 10):
        if not i in base:
            mid = middle(i, base[-1])
            if mid is None or mid in base:
                tmp.append(i)
    return tmp


def generate(base):
    for n in next(base):
        s = base + [n]
        generate(s)
        Seq.append(s)


def count():
    generate([])
    for seq in Seq:
        if len(seq) >= min_seq_length:
            length[len(seq)] += 1

    print "length:", length, "total:", sum(length)


if __name__ == "__main__":
    count()

File: pattern/pattern.go
package pattern

import "fmt"

var (
	sequences [][]int
	min       = 4
)

// return the node between a and b if there is one, or 0.
func middle(a, b int) int {
	if (a+b)%2 != 0 {
		return 0
	}
	if mid := (a + b) / 2; (mid == 5) || (a%3 == b%3) || ((a-1)/3 == (b-1)/3) {
		return mid
	}
	return 0
}

// generate valid moves j+1 given a sequence of moves 1..j.
func next(base []int) []int {
	x := []int{1, 2, 3, 4, 5, 6, 7, 8, 9}
	if n := len(base); n >= 9 {
		return nil
	} else if n == 0 {
		return x
	}
	tmp := []int{}
	for _, i := range x {
		if !contains(i, base) {
			if mid := middle(i, base[len(base)-1]); mid == 0 || contains(mid, base) {
				tmp = append(tmp, i)
			}
		}
	}
	return tmp
}

// check if array contains a number or not
func contains(i int, items []int) bool {
	for _, x := range items {
		if i == x {
			return true
		}
	}
	return false
}

// generate valid sequences recursively
func generate(base []int) {
	for _, n := range next(base) {
		s := append(base, n)
		sequences = append(sequences, s)
		generate(s)
	}
}

// calculate sum of an array
func sum(arr []int) (s int) {
	s = 0
	for _, x := range arr {
		s += x
	}
	return
}

// count all valid sequence by length and calculate sum of them
func count() (lengths []int, total int) {
	sequences = make([][]int, 0)
	generate([]int{})
	lengths = []int{0, 0, 0, 0, 0, 0, 0, 0, 0, 0}
	for _, s := range sequences {
		if len(s) >= min {
			lengths[len(s)]++
		}
	}
	fmt.Println(
		"length:", lengths,
		"total:", sum(lengths))
	return lengths, sum(lengths)
}
